drivers/spi/spi_silabs_eusart.c

```diff
@@ -152,6 +212,314 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+{
+	eusart->CMD_SET = EUSART_CMD_CLEARTX;
+
+	while (EUSART_StatusGet(eusart) & EUSART_STATUS_RXFL) {
+		(void)eusart->RXDATA;
+	}
+
+	while (eusart->STATUS & EUSART_STATUS_CLEARTXBUSY) {
+	}
+}
+
+static inline size_t spi_silabs_longuest_transfer_size(struct spi_context *instance_ctx)
+{
+	uint32_t tx_transfer_size = spi_context_total_tx_len(instance_ctx);
+	uint32_t rx_transfer_size = spi_context_total_rx_len(instance_ctx);
+
+	return MAX(tx_transfer_size, rx_transfer_size);
+}
+
+static uint32_t
+spi_silabs_build_dma_transfer_block(const struct spi_silabs_eusart_config *instance_config,
+				    struct dma_block_config *new_blk_cfg, uint8_t *buffer,
+				    size_t requested_transaction_size, bool is_tx)
+{
+	memset(new_blk_cfg, 0, sizeof(*new_blk_cfg));
+	size_t remaining_transaction_size = 0U;
+
+	/* Set-up source and destination address with increment behavior. */
+	if (is_tx) {
+		new_blk_cfg->dest_address = (uint32_t)&instance_config->base->TXDATA;
+		new_blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		if (buffer) {
+			new_blk_cfg->source_address = (uint32_t)buffer;
+			new_blk_cfg->source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			/* Null buffer pointer means sending dummy byte. */
+			new_blk_cfg->source_address = (uint32_t)&dummy_byte;
+			new_blk_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	} else {
+		new_blk_cfg->source_address = (uint32_t)&instance_config->base->RXDATA;
+		new_blk_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		if (buffer) {
+			new_blk_cfg->dest_address = (uint32_t)buffer;
+			new_blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			/* Null buffer pointer means rx to null byte. */
+			new_blk_cfg->dest_address = (uint32_t)&empty_buffer;
+			new_blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+
+	/* Setup max transfer according to requested transaction size.
+	 * Will top if bigger than the maximum transfer size.
+	 */
+	if (requested_transaction_size <= SPI_DMA_MAX_DESCRIPTOR_TRANSFER_SIZE) {
+		new_blk_cfg->block_size = requested_transaction_size;
+		remaining_transaction_size = 0U;
+	} else {
+		new_blk_cfg->block_size = SPI_DMA_MAX_DESCRIPTOR_TRANSFER_SIZE;
+		remaining_transaction_size = requested_transaction_size - new_blk_cfg->block_size;
+	}
+
+	/* Next block ptr. */
+	if (remaining_transaction_size > 0U) {
+		new_blk_cfg->next_block = &(new_blk_cfg[1]);
+	} else {
+		new_blk_cfg->next_block = NULL;
+	}
+
+	/* Return how much of the requested transaction is still required to be processed. */
+	return remaining_transaction_size;
+}
+
+static void spi_silabs_build_dma_config(const struct device *dev,
+					struct dma_config *current_dma_config,
+					const struct dma_block_config *blk_cfg,
+					uint32_t block_count, bool is_tx)
+{
+	memset(current_dma_config, 0, sizeof(*current_dma_config));
+
+	if (is_tx) {
+		/* Memory to peripheral. */
+		current_dma_config->channel_direction = 0x02U;
```

Comment:
Should this be `enum dma_channel_direction` instead of a magic number?
Though: `MEMORY_TO_PERIPHERAL` is 1 and `PERIPHERAL_TO_MEMORY` is 2, so this seems backwards?